include/glow/Optimizer/GraphOptimizer/NodeSplitting.h

```diff
@@ -0,0 +1,283 @@
+      : SplitNodeOption(splitDims), numChunks_(numChunks),
+        bigChunksFirst_(bigChunksFirst) {
+    CHECK_EQ(splitDims.size(), numChunks.size())
+        << "Mismatch between 'splitDims' and 'numChunks' array sizes!";
+  }
+
+  /// Split a given dimension size \p dimSize along the dimension \p dim.
+  /// \returns the chunks sizes after splitting.
+  std::vector<dim_t> splitAlongDim(size_t dim, dim_t dimSize) const override;
+};
+
+/// Split option used for splitting a node along the dimensions \ref splitDims
+/// with the given chunk sizes \ref chunkSizes for each dimension such that each
+/// chunk obtained during splitting has a size along a given split dimension
+/// at most equal to the given chunk size along that respective dimension.
+/// Since the split does not always result in equal chunks, you can choose to
+/// start the splitting with the bigger chunks first by using the flag
+/// \p bigChunksFirst. The chunk size for each dimension must be lower than the
+/// respective dimension size, otherwise error is thrown.
+class SplitNodeByChunkSize : public SplitNodeOption {
+
+  /// Chunk size for each split dimension.
+  std::vector<dim_t> chunkSizes_;
+
+  /// Whether to start splitting with bigger chunks first.
+  bool bigChunksFirst_{true};
+
+public:
+  /// Ctor.
+  SplitNodeByChunkSize(llvm::ArrayRef<size_t> splitDims,
+                       llvm::ArrayRef<dim_t> chunkSizes,
+                       bool bigChunksFirst = true)
+      : SplitNodeOption(splitDims), chunkSizes_(chunkSizes),
+        bigChunksFirst_(bigChunksFirst) {
+    CHECK_EQ(splitDims.size(), chunkSizes.size())
+        << "Mismatch between 'splitDims' and 'chunkSizes' array sizes!";
+  }
+
+  /// Split a given dimension size \p dimSize along the dimension \p dim.
+  /// \returns the chunks sizes after splitting.
+  std::vector<dim_t> splitAlongDim(size_t dim, dim_t dimSize) const override;
+};
+
+/// Split option used for splitting a node along the dimensions \ref splitDims
+/// with the exact array of chunk sizes \ref chunkSizes for each dimension.
+/// The chunk sizes must be strictly positive and the sum of the chunk sizes
+/// for each dimension must be equal to the respective dimension size, otherwise
+/// error is thrown.
+class SplitNodeByChunkSizes : public SplitNodeOption {
+
+  /// Array of chunk sizes for each split dimension.
+  std::vector<std::vector<dim_t>> chunkSizes_;
+
+public:
+  /// Ctor.
+  SplitNodeByChunkSizes(llvm::ArrayRef<size_t> splitDims,
+                        llvm::ArrayRef<std::vector<dim_t>> chunkSizes)
+      : SplitNodeOption(splitDims), chunkSizes_(chunkSizes) {
+    CHECK_EQ(splitDims.size(), chunkSizes.size())
+        << "Mismatch between 'splitDims' and 'chunkSizes' array sizes!";
+  }
+
+  /// Split a given dimension size \p dimSize along the dimension \p dim.
+  /// \returns the chunks sizes after splitting.
+  std::vector<dim_t> splitAlongDim(size_t dim, dim_t dimSize) const override;
```

Comment:
Should the return type be a mutable vector? If not, return an ArrayRef.

Reply:
Mutable vector so I left it `std::vector`.